### PR TITLE
[test] Remove install from repo.anaconda.com for PT 1.11 (E3) images

### PR DESCRIPTION
**Pull request by kevinyang8**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/dlc_tests/sanity/test_anaconda.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    container_name = test_utils.get_container_name("anaconda", image)
+    test_utils.start_container(container_name, image, ctx)
+
+    # First check to see if image has conda installed, if not, skip test since no packages installed from conda present
+    conda_present = test_utils.run_cmd_on_container(container_name, ctx, "find . -name conda").stdout.strip()
+    if not conda_present:
+        pytest.skip("Image does not have conda installed, skipping test.")
+    
+    # Commands are split in 2 because if warn=True, then even if first command fails silently, no error is raised
+    conda_list_output = test_utils.run_cmd_on_container(container_name, ctx, "conda list --explicit > repo_list.txt", warn=True).stdout.strip()
```
Comment (arjkesh) [inline test/dlc_tests/sanity/test_anaconda.py:21]: this doesn't need to be warn=True right - and you don't really need to check the output afaik. Why not just let it fail if it fails and then see the stack trace?

Comment (kevinyang8) [reply test/dlc_tests/sanity/test_anaconda.py]: you're right, updated

File: pytorch/inference/docker/1.11/py3/cu113/Dockerfile.gpu [hunk 2]
```diff
@@ -34,7 +34,6 @@ ARG CUBLAS_VERSION=11.5.1.109
 ARG OMPI_VERSION=4.1.1
 ARG TS_VERSION=0.5.2
 
-ARG CONDA_PREFIX=/opt/conda
```
Comment (arjkesh) [inline pytorch/inference/docker/1.11/py3/cu113/Dockerfile.gpu:37]: why remove this? isn't the prefix still the same?

comment applies to other dockerfiles as well

File: pytorch/inference/docker/1.11/py3/cu113/Dockerfile.gpu [hunk 3]
```diff
@@ -145,42 +144,37 @@ RUN apt-get update \
  && rm -rf /var/lib/apt/lists/* \
  && apt-get clean
 
-RUN curl -L -o ~/miniconda.sh https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh \
- && chmod +x ~/miniconda.sh \
- && ~/miniconda.sh -b -p ${CONDA_PREFIX} \
- && rm ~/miniconda.sh \
- && ${CONDA_PREFIX}/bin/conda update conda \
- && ${CONDA_PREFIX}/bin/conda install -c conda-forge \
+RUN curl -L -o ~/mambaforge.sh https://github.com/conda-forge/miniforge/releases/download/4.12.0-0/Mambaforge-4.12.0-0-Linux-x86_64.sh \
```
Comment (arjkesh) [inline pytorch/inference/docker/1.11/py3/cu113/Dockerfile.gpu:147]: is there a reason this is tied to this specific version? this seems like a slight change in behavior from the prior experience - since we were previously installing "latest" every time.

File: test/dlc_tests/sanity/test_anaconda.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+
+@pytest.mark.usefixtures("sagemaker")
+@pytest.mark.model("N/A")
+@pytest.mark.integration("anaconda_removal")
+def test_repo_anaconda_not_present(image):
+    """ Test to see if all packages installed in the image do not come from repo.anaconda.com """
+    ctx = Context()
+    container_name = test_utils.get_container_name("anaconda", image)
+    test_utils.start_container(container_name, image, ctx)
```
Comment (arjkesh) [inline test/dlc_tests/sanity/test_anaconda.py:13]: nit: I believe there is a stop_and_remove_container function in test utils as well - you can add this to the end of the function or as a [pytest finalizer](https://docs.pytest.org/en/6.2.x/fixture.html#adding-finalizers-directly). This will just ensure that the containers don't persist on the codebuild host and potentially contribute to memory issues when running sanity tests

File: test/dlc_tests/sanity/test_anaconda.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+    # First check to see if image has conda installed, if not, skip test since no packages installed from conda present
+    conda_present = test_utils.run_cmd_on_container(container_name, ctx, "find . -name conda").stdout.strip()
+    if not conda_present:
+        pytest.skip("Image does not have conda installed, skipping test.")
```
Comment (arjkesh) [inline test/dlc_tests/sanity/test_anaconda.py:18]: nit: helpful to include which image is being tested, somethignlike

```
pytest.skip(f"Image {image} does not have conda installed, skipping test.")
```

In this case you'll want to stop and remove the container, esp since we're not even using it to test anything

File: test/dlc_tests/sanity/test_anaconda.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    
+    # Commands are split in 2 because if warn=True, then even if first command fails silently, no error is raised
+    test_utils.run_cmd_on_container(container_name, ctx, "conda list --explicit > repo_list.txt")
+
+    grep_result = test_utils.run_cmd_on_container(container_name, ctx, "grep repo.anaconda.com repo_list.txt", warn=True).stdout.strip()
+    if grep_result:
+        raise RuntimeError(f"Image contains packages installed from repo.anaconda.com: {grep_result}")
```
[the remark this answers is not in the document]
Comment (kevinyang8) [reply test/dlc_tests/sanity/test_anaconda.py:25]: Ran and file reformatted. I had versioning problems with the black version in `requirements.txt` so I updated the black version as well to 22.3.0 as online issues say this version resolves the problem I was facing.

File: pytorch/inference/docker/1.11/py3/cu113/Dockerfile.gpu (hunk 2 above)
Comment (kevinyang8) [reply pytorch/inference/docker/1.11/py3/cu113/Dockerfile.gpu:37]: I discovered a niche issue when installing using an env var set for CONDA_PREFIX for installing mamba that would cause installation to fail. For consistency, I removed all env vars set for CONDA_PREFIX as well as args set for CONDA_PREFIX.

File: pytorch/inference/docker/1.11/py3/cu113/Dockerfile.gpu (hunk 3 above)
Comment (kevinyang8) [reply pytorch/inference/docker/1.11/py3/cu113/Dockerfile.gpu:147]: This is a good point, I will update to pull latest mamba version from Github.

File: test/dlc_tests/sanity/test_anaconda.py (hunk 4 above)
Comment (kevinyang8) [reply test/dlc_tests/sanity/test_anaconda.py:13]: update to use a try finally statement

File: test/dlc_tests/sanity/test_anaconda.py (hunk 5 above)
Comment (kevinyang8) [reply test/dlc_tests/sanity/test_anaconda.py:18]: updated

File: pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu [hunk 7]
```diff
@@ -145,42 +145,37 @@ RUN apt-get update \
...
- && ${CONDA_PREFIX}/bin/conda install -y \
-    # conda 4.10.0 requires ruamel_yaml to be installed. Currently pinned at latest.
-    ruamel_yaml==0.15.100 \
- && ${CONDA_PREFIX}/bin/conda install -y -c anaconda \
     cython \
-    mkl \
+    mkl \ 
     mkl-include \
-    parso \
+    parso \
```
Comment (saimidu) [inline pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu:158]: nit: Please remove trailing whitespace
```suggestion
    parso \
```

File: pytorch/training/docker/1.11/py3/Dockerfile.cpu [hunk 8]
```diff
@@ -110,27 +111,26 @@ RUN rm -rf /root/.ssh/ \
...
+ && ~/mambaforge.sh -b -p /opt/conda \
+ && rm ~/mambaforge.sh \
  && /opt/conda/bin/conda update conda \
  && /opt/conda/bin/conda install -c conda-forge \
     python=$PYTHON_VERSION \
- && /opt/conda/bin/conda install -y -c anaconda \
-    # conda 4.10.0 requires ruamel_yaml to be installed. Currently pinned at latest.
     cython \
-    mkl \
+    mkl \
```
Comment (saimidu) [inline pytorch/training/docker/1.11/py3/Dockerfile.cpu:122]: nit: Please remove trailing whitespace.
```suggestion
    mkl \
```

File: pytorch/training/docker/1.11/py3/Dockerfile.cpu [hunk 9]
```diff
@@ -110,27 +111,26 @@ RUN rm -rf /root/.ssh/ \
...
     cython \
-    mkl \
+    mkl \ 
     mkl-include \
-    "pyopenssl>=17.5.0" \
-    ruamel_yaml==0.15.100 \
-    typing \
- && conda install -y \
+    parso \ 
+    typing \
```
Comment (saimidu) [inline pytorch/training/docker/1.11/py3/Dockerfile.cpu:125]: nit: Please remove trailing whitespace
```suggestion
    parso \
    typing \
```

File: pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu [hunk 10]
```diff
@@ -145,42 +145,37 @@ RUN apt-get update \
...
     cython \
-    mkl \
+    mkl \ 
     mkl-include \
-    parso \
+    parso \ 
     scipy \
-    typing \
- && ${CONDA_PREFIX}/bin/conda clean -ya
+    typing \
```
Comment (saimidu) [inline pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu:160]: nit: Please remove trailing whitespace
```suggestion
    typing \
```

File: pytorch/training/docker/1.11/py3/cu113/Dockerfile.gpu [hunk 11]
```diff
@@ -1,6 +1,7 @@
 ARG PYTHON=python3
 ARG PYTHON_VERSION=3.8.12
 ARG PYTHON_SHORT_VERSION=3.8
+ARG MAMBA_VERSION=4.12.0-0
```
Comment (saimidu) [inline pytorch/training/docker/1.11/py3/cu113/Dockerfile.gpu:4]: nit: Please make the `MAMBA_VERSION` ARG definition consistent across both the CPU and the GPU dockerfiles, by either defining it globally, or by defining it specifically in the `common` section.

File: test/dlc_tests/sanity/test_anaconda.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+            pytest.skip(f"Image {image} does not have conda installed, skipping test.")
+
+        # Commands are split in 2 because if warn=True, then even if first command fails silently, no error is raised
+        test_utils.run_cmd_on_container(container_name, ctx, "conda list --explicit > repo_list.txt")
+
+        grep_result = test_utils.run_cmd_on_container(
+            container_name, ctx, "grep repo.anaconda.com repo_list.txt", warn=True
+        ).stdout.strip()
+        if grep_result:
+            raise RuntimeError(f"Image {image} contains packages installed from repo.anaconda.com: {grep_result}")
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_anaconda.py:28]: Please add a statement about the follow-up action that we expect the person looking at these error logs to take.

Something like "Please ensure that these packages are obtained through conda-forge or other alternatives".

Comment (saimidu) [reply test/dlc_tests/sanity/test_anaconda.py:28]: Also, I don't think this test requires a separate test script of its own. Why not put this test in `test_pre_release.py`?

File: test/dlc_tests/sanity/test_anaconda.py [hunk 13]
```diff
@@ -25,7 +25,9 @@ def test_repo_anaconda_not_present(image):
             container_name, ctx, "grep repo.anaconda.com repo_list.txt", warn=True
         ).stdout.strip()
         if grep_result:
-            raise RuntimeError(f"Image {image} contains packages installed from repo.anaconda.com: {grep_result}")
+            raise RuntimeError(
+                f"Image {image} contains packages installed from repo.anaconda.com. \
+                Please ensure that these packages are obtained through conda-forge or other alternatives: {grep_result}"
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_anaconda.py:30]: Please change this to
```suggestion
                f"Image {image} contains packages installed from repo.anaconda.com. "
                f"Please ensure that these packages are obtained through conda-forge or other alternatives: {grep_result}"
```

**Review by saimidu — DISMISSED** [review]

**Review by saimidu — APPROVED** [review]
